File: bukkit/src/main/java/io/github/rothes/protocolstringreplacer/packetlistener/server/itemstack/WindowItemsPost11.java
package io.github.rothes.protocolstringreplacer.packetlistener.server.itemstack;

import com.comphenix.protocol.PacketType;
import com.comphenix.protocol.events.PacketEvent;
import com.comphenix.protocol.reflect.StructureModifier;
import io.github.rothes.protocolstringreplacer.ProtocolStringReplacer;
import io.github.rothes.protocolstringreplacer.api.replacer.ReplacerConfig;
import io.github.rothes.protocolstringreplacer.api.user.PsrUser;
import io.github.rothes.protocolstringreplacer.replacer.ReplacerManager;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public final class WindowItemsPost11 extends AbstractServerItemPacketListener {

    public WindowItemsPost11() {
        super(PacketType.Play.Server.WINDOW_ITEMS);
    }

    protected void process(@NotNull PacketEvent packetEvent) {
        PsrUser user = getEventUser(packetEvent);
        if (user == null) {
            return;
        }
        user.clearUserMetaCache();
        ReplacerManager replacerManager = ProtocolStringReplacer.getInstance().getReplacerManager();
        List<ReplacerConfig> nbt = replacerManager.getAcceptedReplacers(user, itemNbtFilter);
        List<ReplacerConfig> display = replacerManager.getAcceptedReplacers(user, itemDisplayFilter);
        List<ReplacerConfig> entries = replacerManager.getAcceptedReplacers(user, itemEntriesFilter);

        StructureModifier<List<ItemStack>> itemListModifier = packetEvent.getPacket().getItemListModifier();
        List<ItemStack> read = itemListModifier.read(0);
        boolean saveMeta = !user.isInAnvil();
        for (ItemStack itemStack : read) {
            if (itemStack.getType() == Material.AIR) {
                saveMeta = true;
                continue;
            }
            boolean blocked = replaceItemStack(packetEvent, user, listenType, itemStack, nbt, display, entries,
                    // Avoid too many packets kick
                    saveMeta);
            saveMeta = true;
            if (blocked) {
                return;
            }
        }
        itemListModifier.write(0, read);
    }

}

File: bukkit/src/main/kotlin/io/github/rothes/protocolstringreplacer/packetlistener/server/itemstack/WindowItemsPost11.kt
package io.github.rothes.protocolstringreplacer.packetlistener.server.itemstack

import com.comphenix.protocol.PacketType
import com.comphenix.protocol.events.PacketEvent
import io.github.rothes.protocolstringreplacer.ProtocolStringReplacer
import io.github.rothes.protocolstringreplacer.get
import io.github.rothes.protocolstringreplacer.set
import org.bukkit.Material

class WindowItemsPost11 : BaseServerItemPacketListener(PacketType.Play.Server.WINDOW_ITEMS) {

    override fun process(packetEvent: PacketEvent) {
        val user = getEventUser(packetEvent) ?: return
        user.clearUserItemRestoreCache()
        val replacerManager = ProtocolStringReplacer.getInstance().replacerManager
        val nbt = replacerManager.getAcceptedReplacers(user, itemNbtFilter)
        val lore = replacerManager.getAcceptedReplacers(user, itemLoreFilter)
        val entries = replacerManager.getAcceptedReplacers(user, itemEntriesFilter)

        val itemListModifier = packetEvent.packet.itemListModifier
        var saveMeta = !user.isInAnvil
        itemListModifier[0] = itemListModifier[0].map { itemStack ->
            if (itemStack.type == Material.AIR) {
                saveMeta = true
                return@map itemStack
            }
            replaceItemStack(packetEvent, user, listenType, itemStack, nbt, lore, entries, saveMeta).also {
                saveMeta = true
            } ?: return
        }
        val itemModifier = packetEvent.packet.itemModifier
        if (itemModifier.size() != 0) {
            // Since 1.17.1
            val itemStack = itemModifier[0]
            if (itemStack.type == Material.AIR) {
                return
            }
            itemModifier[0] = replaceItemStack(packetEvent, user, listenType, itemStack, nbt, lore, entries, true) ?: return
        }
    }
}
